### PR TITLE
*: Added leaktest to all unit tests and fixed all found leaks

### PR DESCRIPTION
**Pull request by Bplotka**
- Fixed cluster leaks.

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by Bplotka** [comment]
https://github.com/improbable-eng/thanos/issues/232

**Comment by Bplotka** [comment]
TODO in next iteration -> https://github.com/improbable-eng/thanos/issues/234 (not so critical)

File: Gopkg.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 [[constraint]]
   name = "cloud.google.com/go"
-  version = "0.16.0"
+  version = "0.19.0"
```
Comment (mattbostock) [inline Gopkg.toml]: Is the version upgrade related to fixing leaks? If so, what's the upstream fix/issue?

If not related to fixing the leaks, would be better to use a separate commit.

File: pkg/testutil/objstore.go [hunk 2]
```diff
@@ -21,29 +21,35 @@ import (
 //
 // TODO(fabxc): define object storage interface and have this method return
 // mocks or actual remote buckets depending on env vars.
-func NewObjectStoreBucket(t testing.TB) (*gcs.Bucket, func()) {
+func NewObjectStoreBucket(t testing.TB) (*gcs.Bucket, func(), error) {
```
Comment (mattbostock) [inline pkg/testutil/objstore.go]: What's the reason for this change? It seems to add a lot of boiler plate to the tests.

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -222,10 +218,15 @@ func (p *Peer) SetTimestamps(mint int64, maxt int64) {
...
 }
 
-// Leave the cluster, waiting up to timeout.
-func (p *Peer) Leave(timeout time.Duration) error {
+// Close leaves the cluster waiting up to timeout and shutdowns peer if cluster left.
+// TODO(bplotka): Add this method into run.Group closing logic for each command. This will improve graceful shutdown.
+func (p *Peer) Close(timeout time.Duration) error {
+	err := p.mlist.Leave(timeout)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply pkg/cluster/cluster.go]: I think that is the correct behavior isn't it? If leave fails, member is still in cluster.

[the remark this answers is not in the document]
Comment (Bplotka) [reply pkg/cluster/cluster.go]: We can do your option as well, but then there is no point in returning error I think

File: pkg/cluster/cluster_test.go [hunk 4]
```diff
@@ -62,11 +61,15 @@ func sortStr(str []string) []string {
 }
 
 func TestPeers_PropagatingState(t *testing.T) {
+	defer leaktest.CheckTimeout(t, 10*time.Second)()
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply pkg/cluster/cluster_test.go]: I twas too short for my experience in previous projects - yes. I think there is no  harm in it -> CI tends to be slow from time to time.

File: Gopkg.toml (hunk 1 above)
Comment (Bplotka) [reply Gopkg.toml]: True. It was rather my attempt for fixing https://github.com/improbable-eng/thanos/issues/234 ;p but seems that new version has leak as well (: Can revert this.

File: pkg/cluster/cluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattbostock) [reply pkg/cluster/cluster.go]: > I think that is the correct behavior isn't it? If leave fails, member is still in cluster.

I wondered that, but I'm not sure that's guaranteed either. The only reason I found for `Leave()` to return an error (in the current implementation) is due to a timeout. Which is why I think proceeding with the shutdown on error might be preferable.

> We can do your option as well, but then there is no point in returning error I think

Agreed. If you have access to a logger, it might be nice to log the error returned by `Leave()`, but I think we should continue to `Shutdown()`.

File: pkg/testutil/objstore.go (hunk 2 above)
Comment (Bplotka) [reply pkg/testutil/objstore.go]: I would rather prefer boilerplate, than being unable to cleanly return. It helps to have clear "cleanup path". If we use "cool" `testuitl.Ok(t, err)` there is no way to cleanly close/cleanup something that was allocated already. 

Otherwise we can handle this by just using `t.FailNow` + `fmt.Print` but it is just worse then `return nil, err`

And IMO having
 ```
bkt := testutilNewObjectStoreBucket
```
vs
```
bkt, err := testutilNewObjectStoreBucket
testuitl.Ok(t, err)
```
is really same. Not much of boilerplate in exchange for less prone to bug code. What do you think?

Comment (mattbostock) [reply pkg/testutil/objstore.go]: Thanks for explaining, that makes sense.

File: pkg/cluster/cluster_test.go [hunk 5]
```diff
@@ -1,20 +1,19 @@
...
 	"testing"
 	"time"
 
-	"context"
-	"errors"
-	"sort"
-
-	"reflect"
-
+	"github.com/fortytw2/leaktest"
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply pkg/cluster/cluster_test.go]: yea, this package saved our as**s a lot of times.

**Comment by Bplotka** [comment]
Is that ok to be merged @mattbostock ?

**Comment by mattbostock** [comment]
LGTM 👍
